web/auth/auth.go

```diff
@@ -177,145 +144,148 @@ func sessionData(session *sessions.Session) string {
+		fmt.Println("AUTH2LOGIN started with: ", r.Method) // tmp
+		if r.Method != "GET" {
+			authenticationError(logger, w, fmt.Sprintf("illegal request method: %s", r.Method))
+			return
+		}
 
-		provider, err := gothic.GetProviderName(r)
+		// Start or refresh the session.
+		// Issue: server generates a new random key to encode sessions on each restart.
+		// A session from before the restart will be detected, but cannot be decoded with
+		// the new key, resulting in an error. Instead of returning, we attempt to refresh the session first.
+		s, err := sessionStore.Get(r, SessionKey)
 		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
-		}
-		var teacher int
-		if strings.HasSuffix(provider, TeacherSuffix) {
-			teacher = 1
+			if err := sessionStore.Save(r, w, s); err != nil {
+				authenticationError(logger, w, fmt.Sprintf("failed to create new session (%s): %s", SessionKey, err))
+				return
+			}
 		}
-		logger.Debugf("Provider: %v ; Teacher: %v", provider, teacher)
 
-		qv := r.URL.Query()
-		logger.Debugf("qv: %v", qv)
-		redirect := extractRedirectURL(r, Redirect)
-		logger.Debugf("redirect: %v", redirect)
-		// TODO: Add a random string to protect against CSRF.
-		qv.Set(State, strconv.Itoa(teacher)+redirect)
-		logger.Debugf("State: %v", strconv.Itoa(teacher)+redirect)
-		r.URL.RawQuery = qv.Encode()
-		logger.Debugf("RawQuery: %v", r.URL.RawQuery)
+		// Get provider name.
+		provider := GetProviderName(r.URL.Path, 2)
+		if provider == "" {
+			authenticationError(logger, w, "incorrect request URL")
+			return
+		}
+		providerConfig, ok := authConfig.providers[provider]
+		if !ok {
+			authenticationError(logger, w, "provider is not enabled")
+			return
+		}
 
-		url, err := gothic.GetAuthURL(w, r)
-		if err != nil {
-			logger.Error(err.Error())
-			return echo.NewHTTPError(http.StatusBadRequest, err.Error())
+		// Check teacher suffix, update scopes if teacher.
+		// Won't be necessary with GitHub App.
+		if strings.Contains(r.URL.Path, TeacherSuffix) {
+			logger.Debug("found teacher suffix on login")
+			providerConfig.Scopes = teacherScopes
+		} else {
+			providerConfig.Scopes = studentScopes
 		}
-		logger.Debugf("Redirecting to %s to perform authentication; AuthURL: %v", provider, url)
-		return c.Redirect(http.StatusTemporaryRedirect, url)
+		logger.Debugf("Provider callback URL: %s", providerConfig.RedirectURL)
+		redirectURL := providerConfig.AuthCodeURL(secret)
+		logger.Debugf("redirecting to AuthURL: %v", redirectURL)
+		http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
 	}
 }
 
 // OAuth2Callback handles the callback from an oauth2 provider.
-func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, scms *Scms) echo.HandlerFunc {
-	return func(c echo.Context) error {
+func OAuth2Callback(logger *zap.SugaredLogger, db database.Database, authConfig *AuthConfig, scms *Scms, secret string) http.HandlerFunc {
```

Comment:
This func is 169 lines long. There are many opportunities to extract code snippets into functions or methods. In particular, one or more methods could be added to the `AuthConfig` type.